lib/improver/lapse_rate.py

```diff
@@ -0,0 +1,308 @@
+                Cube containing orography data (metres)
+
+            land_sea_mask_cube (iris.cube.Cube):
+                Cube containing a binary land-sea mask. True for land-points
+                and False for Sea.
+
+        Keyword Args:
+            max_height_diff (float):
+                Maximum allowable height difference between the central point
+                and its neighboours (metres).
+
+            footprint_len (int):
+                Size of neighbourhood around each point.
+
+        Returns:
+            lapse_rate_cube (iris.cube.Cube):
+                Cube containing lapse rate (Km-1)
+
+        Raises
+        ------
+        TypeError: If input cubes are not cubes
+
+        """
+
+        class fnc_class:
+            def __init__(self, allbuffers):
+                # Initialises the iterator:
+                self.i = 0
+                # Saves all the buffers into this array.
+                self.allbuffers = allbuffers
+
+            def filter(self, buffer):
+                self.allbuffers[self.i, :] = buffer
+                self.i += 1
+                return 0.0  # Need to return a value
+
+        def calc_lapse_rate(temperature, orography, ind_central_point):
+            """
+            Least-squares fit to local temperature and altitude data to
+            calculate a local lapse rate.
+
+            """
+
+            # If central point NaN then return blank value.
+            if np.isnan(temperature[ind_central_point]):
+                return 0.0
+
+            # Remove NaN values before calculation.
+            y_data = temperature[~np.isnan(temperature)]
+            x_data = orography[~np.isnan(orography)]
+
+            # If empty array - return blank value.
+            if len(y_data) == 0:
+                return 0.0
+
+            matrix = np.stack([x_data, np.ones(len(x_data))], axis=0).T
+            gradient, intercept = lstsq(matrix, y_data)[0]
+
+            return gradient
+
+        def create_height_diff_mask(all_orog_subsections, ind_central_point,
+                                    max_height_diff):
+            """
+            Create a mask for any neighbouring points where the height
+            difference from the central point is greater than max_height_diff.
+
+            Slice through the orography subsection array to remove central
+            points. Then subtract central height value from neighbourhoods.
+            """
+            central_points = all_orog_subsections[:, ind_central_point]
+            central_points = np.swapaxes([central_points], 0, 1)
+
+            height_diff = np.subtract(all_orog_subsections, central_points)
+            height_diff = np.absolute(height_diff)
+
+            height_diff_mask = np.where(height_diff >= max_height_diff, True,
+                                        False)
+            return height_diff_mask
+
+        DALR = 0.0098  # Dry Adiabatic Lapse Rate in unit of K/m-1
```

Comment:
This should be added to ```constants.py``` with units. This code could then work with a more general set of units if so desired.